src/main/java/io/xdb/core/state/AsyncState.java

```diff
@@ -35,16 +37,17 @@ default CommandRequest waitUntil(final I input) {
      * It's called after the commands specified in {@link AsyncState#waitUntil} have been completed or, in the case where {@link AsyncState#waitUntil} is skipped, it is invoked directly.
      *
      * @param input
+     * @param request
      * @return
      */
-    StateDecision execute(final I input);
+    StateDecision execute(final I input, final Communication communication, final AsyncStateExecuteRequest request);
```

Comment:
It seems direcly put the `request` here will make the API look concise.

Reply:
This `AsyncStateExecuteRequest` is an internal structure of the protocol that users don't really have to understand. E.g. 
1. we need to decoded the payload for them, 
2. re-organize the commands
3. There are a lot more information than just this localQ command results, it will contain like global&local attribute values. 

Essentially, the whole execute API is prepared by this request. Users will have a very hard time to understand this --- essentially, they will be doing like what we wrote integ test in iWF server if you expose this structure

Reply:
True. The original thought from me is that in iwf java sdk, there are too many parameters in the waitUntil and execute methods. So I would like to reduce the parameter count for both methods. I just created AsyncStateExecuteFeatures and AsyncStateWaitUntilFeatures (is features a good name?) classes to wrap everything except the input.

Reply:
yeah agreed there are too many. Naming/API structure is really, really hard, and I am still open to better ideas. 

Actually I got to the current states actually after many iterations Lol. This is one of the early version: https://docs.google.com/document/d/15CETNk9ewiP7M_6N9s7jo-Wm57WG977hch9kTVnaExA/edit

Also, in XDB, we will not have persistence in waitUntil API because from iWF users, we learned people don't really need to read/write data attributes in waitUntil. So we will reduce this in XDB. So I am imagine the end state of the state APIs:

```java
public CommandRequest waitUntil(context, input, communication);

public StateDecision execute(context, input, commandResults, persistence, communication);

```

I have tried hard to reduce the number of params too(and adjusted a lot of namings as well). I have thought about putting everything into a single parameter too. 

SDK is for user to write code. So the goal of the SDK API to make it extremely clear and friendly about how is available and what can be done with each API. that's why we grouped to persistence/communication/commandResults. Putting everything under one parameter will reduce readability and less natural to write code for users. 

Also, I think the number of params are fine today when users are using IDE(mostly). I haven't seen anyone complained this in iWF yet.

Reply:
True. Being more direct should reduce less effrots to use these APIs. I will expand the parameters.